File: backend/app/services/clerk_permission_service.py
from app.constants import (
    CLERK_PERMISSIONS,
    DEFAULT_CLERK_PERMISSIONS,
    STATUS_TRANSITIONS_BY_PERMISSION,
)
from app.models.user import User


def normalize_clerk_permissions(perms: list | None) -> list[str]:
    if not perms:
        return list(DEFAULT_CLERK_PERMISSIONS)
    return [p for p in perms if p in CLERK_PERMISSIONS]


def get_clerk_permissions(user: User) -> list[str]:
    if user.role == "admin":
        return list(CLERK_PERMISSIONS)
    if user.role != "clerk":
        return []
    return normalize_clerk_permissions(user.clerk_permissions)


def clerk_has_permission(user: User, permission: str) -> bool:
    if user.role == "admin":
        return True
    if user.role != "clerk":
        return False
    return permission in get_clerk_permissions(user)


def clerk_has_any_permission(user: User, permissions: tuple[str, ...] | list[str]) -> bool:
    if user.role == "admin":
        return True
    if user.role != "clerk":
        return False
    perms = set(get_clerk_permissions(user))
    return any(p in perms for p in permissions)


def assert_status_transition_allowed(user: User, from_status: str, to_status: str) -> None:
    if user.role == "admin":
        return
    if from_status == to_status:
        return

    if user.role != "clerk":
        raise ValueError("Clerk access required")

    allowed = set()
    user_perms = get_clerk_permissions(user)
    for perm in user_perms:
        allowed.update(STATUS_TRANSITIONS_BY_PERMISSION.get(perm, set()))

    if (from_status, to_status) not in allowed:
        raise ValueError(
            f"You are not allowed to change status from {from_status} to {to_status}"
        )


def assert_target_status_allowed(user: User, to_status: str, from_status: str | None = None) -> None:
    """For bulk updates when from_status may vary per package."""
    if user.role == "admin":
        return
    if from_status is not None:
        assert_status_transition_allowed(user, from_status, to_status)
        return

    if user.role != "clerk":
        raise ValueError("Clerk access required")

    user_perms = get_clerk_permissions(user)
    allowed_targets = set()
    for perm in user_perms:
        for _from, to in STATUS_TRANSITIONS_BY_PERMISSION.get(perm, set()):
            allowed_targets.add(to)

    if to_status not in allowed_targets:
        raise ValueError(f"You are not allowed to set status to {to_status}")

File: server/src/services/clerkPermissionService.ts
import {
  CLERK_PERMISSIONS,
  DEFAULT_CLERK_PERMISSIONS,
  STATUS_TRANSITIONS_BY_PERMISSION,
} from '../constants.js'
import type { UserRow } from '../db/schema/index.js'

export function normalizeClerkPermissions(perms: string[] | null | undefined): string[] {
  if (!perms?.length) return [...DEFAULT_CLERK_PERMISSIONS]
  return perms.filter((p) => (CLERK_PERMISSIONS as readonly string[]).includes(p))
}

export function getClerkPermissions(user: UserRow): string[] {
  if (user.role === 'admin') return [...CLERK_PERMISSIONS]
  if (user.role !== 'clerk') return []
  return normalizeClerkPermissions(user.clerkPermissions ?? undefined)
}

export function clerkHasPermission(user: UserRow, permission: string): boolean {
  if (user.role === 'admin') return true
  if (user.role !== 'clerk') return false
  return getClerkPermissions(user).includes(permission)
}

export function clerkHasAnyPermission(user: UserRow, permissions: string[]): boolean {
  if (user.role === 'admin') return true
  if (user.role !== 'clerk') return false
  const perms = new Set(getClerkPermissions(user))
  return permissions.some((p) => perms.has(p))
}

export function assertStatusTransitionAllowed(
  user: UserRow,
  fromStatus: string,
  toStatus: string,
): void {
  if (user.role === 'admin') return
  if (fromStatus === toStatus) return
  if (user.role !== 'clerk') throw new Error('Clerk access required')

  const allowed = new Set<string>()
  for (const perm of getClerkPermissions(user)) {
    for (const transition of STATUS_TRANSITIONS_BY_PERMISSION[perm] ?? []) {
      allowed.add(transition)
    }
  }

  if (!allowed.has(`${fromStatus}->${toStatus}`)) {
    throw new Error(`You are not allowed to change status from ${fromStatus} to ${toStatus}`)
  }
}
